src/concurrency/ocf_metadata_concurrency.c

```diff
@@ -8,46 +8,94 @@
 
 void ocf_metadata_concurrency_init(struct ocf_metadata_lock *metadata_lock)
 {
+	unsigned i;
+
 	env_spinlock_init(&metadata_lock->eviction);
 	env_rwlock_init(&metadata_lock->status);
 	env_rwsem_init(&metadata_lock->global);
+
+	for (i = 0; i < OCF_IO_CLASS_MAX; i++) {
+		env_spinlock_init(&metadata_lock->partition[i]);
+	}
 }
 
 void ocf_metadata_concurrency_deinit(struct ocf_metadata_lock *metadata_lock)
 {
+	unsigned i;
+
+	for (i = 0; i < OCF_IO_CLASS_MAX; i++) {
+		env_spinlock_destroy(&metadata_lock->partition[i]);
+	}
+
 	env_spinlock_destroy(&metadata_lock->eviction);
 	env_rwlock_destroy(&metadata_lock->status);
 	env_rwsem_destroy(&metadata_lock->global);
 }
 
 int ocf_metadata_concurrency_attached_init(
 		struct ocf_metadata_lock *metadata_lock, ocf_cache_t cache,
-		uint64_t hash_table_entries)
+		uint32_t hash_table_entries, uint32_t colision_table_pages)
 {
-	uint64_t i;
+	uint32_t i;
+	int err = 0;
 
-	metadata_lock->cache = cache;
-	metadata_lock->num_hash_entries = hash_table_entries;
 	metadata_lock->hash = env_vzalloc(sizeof(env_rwsem) *
 			hash_table_entries);
-	if (!metadata_lock->hash)
+	metadata_lock->collision_pages = env_vzalloc(sizeof(env_rwsem) *
+			colision_table_pages);
+	if (!metadata_lock->hash ||
+			!metadata_lock->collision_pages) {
+		env_vfree(metadata_lock->hash);
+		env_vfree(metadata_lock->collision_pages);
+		metadata_lock->hash = NULL;
+		metadata_lock->collision_pages = NULL;
 		return -OCF_ERR_NO_MEM;
+	}
 
 	for (i = 0; i < hash_table_entries; i++)
 		env_rwsem_init(&metadata_lock->hash[i]);
+	for (i = 0; i < colision_table_pages; i++) {
+		err = env_rwsem_init(&metadata_lock->collision_pages[i]);
```

Comment:
We want similar error handling in line 56.

Reply:
done